# Workbook

Use this notebook to complete the exercises throughout the workshop.

#### Table of Contents
- [Section 1 &ndash; Getting Started with Pandas](#Section-1)
- [Section 2 &ndash; Data Wrangling](#Section-2)
- [Section 3 &ndash; Data Visualization](#Section-3)

--- 

### Section 1

#### Exercise 1.1
##### Create a DataFrame by reading in the `2019_Yellow_Taxi_Trip_Data.csv` file. Examine the first 5 rows.

In [1]:
import pandas as pd
df = pd.read_csv('../data/2019_Yellow_Taxi_Trip_Data.csv')
df.head(5)

vendorid     tpep_pickup_datetime    tpep_dropoff_datetime   
0         2  2019-10-23T16:39:42.000  2019-10-23T17:14:10.000  \
1         1  2019-10-23T16:32:08.000  2019-10-23T16:45:26.000   
2         2  2019-10-23T16:08:44.000  2019-10-23T16:21:11.000   
3         2  2019-10-23T16:22:44.000  2019-10-23T16:43:26.000   
4         2  2019-10-23T16:45:11.000  2019-10-23T16:58:49.000   

   passenger_count  trip_distance  ratecodeid store_and_fwd_flag   
0                1           7.93           1                  N  \
1                1           2.00           1                  N   
2                1           1.36           1                  N   
3                1           1.00           1                  N   
4                1           1.96           1                  N   

   pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax   
0           138           170             1         29.5    1.0      0.5  \
1            11            26             1         10.5    1.0      0.5   
2           163           162             1          9.5    1.0      0.5   
3           170           163             1         13.0    1.0      0.5   
4           163           236             1         10.5    1.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount   
0        7.98          6.12                    0.3         47.90  \
1        0.00          0.00                    0.3         12.30   
2        2.00          0.00                    0.3         15.80   
3        4.32          0.00                    0.3         21.62   
4        0.50          0.00                    0.3         15.30   

   congestion_surcharge  
0                   2.5  
1                   0.0  
2                   2.5  
3                   2.5  
4                   2.5

#### Exercise 1.2
##### Find the dimensions (number of rows and number of columns) in the data.

In [2]:
print(df.shape)
df.info()
df.describe()

(10000, 18)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   vendorid               10000 non-null  int64  
 1   tpep_pickup_datetime   10000 non-null  object 
 2   tpep_dropoff_datetime  10000 non-null  object 
 3   passenger_count        10000 non-null  int64  
 4   trip_distance          10000 non-null  float64
 5   ratecodeid             10000 non-null  int64  
 6   store_and_fwd_flag     10000 non-null  object 
 7   pulocationid           10000 non-null  int64  
 8   dolocationid           10000 non-null  int64  
 9   payment_type           10000 non-null  int64  
 10  fare_amount            10000 non-null  float64
 11  extra                  10000 non-null  float64
 12  mta_tax                10000 non-null  float64
 13  tip_amount             10000 non-null  float64
 14  tolls_amount           10000 non-null  floa

vendorid  passenger_count  trip_distance    ratecodeid   
count  10000.000000     10000.000000   10000.000000  10000.000000  \
mean       1.633700         1.497700       3.015250      1.084200   
std        0.481817         1.139353       4.148063      0.418244   
min        1.000000         0.000000       0.000000      1.000000   
25%        1.000000         1.000000       0.920000      1.000000   
50%        2.000000         1.000000       1.500000      1.000000   
75%        2.000000         2.000000       2.760000      1.000000   
max        2.000000         6.000000      38.110000      5.000000   

       pulocationid  dolocationid  payment_type   fare_amount         extra   
count  10000.000000  10000.000000  10000.000000  10000.000000  10000.000000  \
mean     166.900400    166.313500      1.301000     15.106313      1.960235   
std       63.791288     68.525953      0.486644     13.954762      1.392940   
min        1.000000      1.000000      1.000000    -52.000000     -4.500000   
25%      132.000000    132.000000      1.000000      7.000000      1.000000   
50%      162.000000    163.000000      1.000000     10.000000      1.000000   
75%      234.000000    236.000000      2.000000     16.000000      3.500000   
max      265.000000    265.000000      4.000000    176.000000      7.000000   

            mta_tax    tip_amount  tolls_amount  improvement_surcharge   
count  10000.000000  10000.000000  10000.000000           10000.000000  \
mean       0.492450      2.634494      0.623447               0.298140   
std        0.071544      3.409800      6.437507               0.032537   
min       -0.500000      0.000000     -6.120000              -0.300000   
25%        0.500000      0.000000      0.000000               0.300000   
50%        0.500000      2.000000      0.000000               0.300000   
75%        0.500000      3.250000      0.000000               0.300000   
max        0.500000     43.000000    612.000000               0.300000   

       total_amount  congestion_surcharge  
count  10000.000000          10000.000000  
mean      22.564659              2.282800  
std       19.209255              0.720946  
min      -65.920000             -2.500000  
25%       12.375000              2.500000  
50%       16.300000              2.500000  
75%       22.880000              2.500000  
max      671.800000              2.750000

#### Exercise 1.3
##### Using the data in the `2019_Yellow_Taxi_Trip_Data.csv` file, calculate summary statistics for the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` columns.

In [3]:
df[['fare_amount','tip_amount', 'tolls_amount', 'total_amount']].describe()

fare_amount    tip_amount  tolls_amount  total_amount
count  10000.000000  10000.000000  10000.000000  10000.000000
mean      15.106313      2.634494      0.623447     22.564659
std       13.954762      3.409800      6.437507     19.209255
min      -52.000000      0.000000     -6.120000    -65.920000
25%        7.000000      0.000000      0.000000     12.375000
50%       10.000000      2.000000      0.000000     16.300000
75%       16.000000      3.250000      0.000000     22.880000
max      176.000000     43.000000    612.000000    671.800000

#### Exercise 1.4
##### Isolate the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` for the longest trip by distance (`trip_distance`).

In [4]:
df.loc[df.trip_distance.idxmax(), ['fare_amount','tip_amount', 'tolls_amount', 'total_amount']]

fare_amount      176.0
tip_amount       18.29
tolls_amount      6.12
total_amount    201.21
Name: 8338, dtype: object

---

### Section 2

#### Exercise 2.1
##### Read in the meteorite data from the `Meteorite_Landings.csv` file, rename the `mass (g)` column to `mass`, and drop all the latitude and longitude columns. Sort the result by mass in descending order.

In [9]:
import pandas as pd
meteorites = pd.read_csv("../data/Meteorite_Landings.csv").rename(columns={'mass (g)' : 'mass'}).sort_values(['mass'], ascending=False)
meteorites.head()

name     id nametype      recclass        mass   fall   
16392             Hoba  11890    Valid     Iron, IVB  60000000.0  Found  \
5373         Cape York   5262    Valid   Iron, IIIAB  58200000.0  Found   
5365   Campo del Cielo   5247    Valid  Iron, IAB-MG  50000000.0  Found   
5370     Canyon Diablo   5257    Valid  Iron, IAB-MG  30000000.0  Found   
3455           Armanty   2335    Valid    Iron, IIIE  28000000.0  Found   

                         year    reclat    reclong             GeoLocation  
16392  01/01/1920 12:00:00 AM -19.58333   17.91667   (-19.58333, 17.91667)  
5373   01/01/1818 12:00:00 AM  76.13333  -64.93333   (76.13333, -64.93333)  
5365   12/22/1575 12:00:00 AM -27.46667  -60.58333  (-27.46667, -60.58333)  
5370   01/01/1891 12:00:00 AM  35.05000 -111.03333     (35.05, -111.03333)  
3455   01/01/1898 12:00:00 AM  47.00000   88.00000            (47.0, 88.0)

#### Exercise 2.2
##### Using the meteorite data from the `Meteorite_Landings.csv` file, update the `year` column to only contain the year, convert it to a numeric data type, and create a new column indicating whether the meteorite was observed falling before 1970. Set the index to the `id` column and extract all the rows with IDs between 10,036 and 10,040 (inclusive) with `loc[]`.

###### **Hint 1**: Use `year.str.slice()` to grab a substring.

###### **Hint 2**: Make sure to sort the index before using `loc[]` to select the range.

###### **Bonus**: There's a data entry error in the `year` column. Can you find it? (Don't spend too much time on this.)

In [23]:
import pandas as pd
meteorites = pd.read_csv("../data/Meteorite_Landings.csv").assign(
    year=lambda x: pd.to_numeric(x.year.str.slice(6, 10)),
    pre_1970=lambda x: (x.fall == 'Fell') & (x.year < 1970)
)
meteorites.set_index('id').sort_index().loc[10_036:10_040]

name nametype         recclass  mass (g)   fall    year    reclat   
id                                                                              
10036     Enigma    Valid               H4      94.0  Found  1967.0  31.33333  \
10037       Enon    Valid  Iron, ungrouped     763.0  Found  1883.0  39.86667   
10038      Enshi    Valid               H5    8000.0   Fell  1974.0  30.30000   
10039  Ensisheim    Valid              LL6  127000.0   Fell  1491.0  47.86667   

         reclong            GeoLocation  pre_1970  
id                                                 
10036  -82.31667  (31.33333, -82.31667)     False  
10037  -83.95000     (39.86667, -83.95)     False  
10038  109.50000          (30.3, 109.5)     False  
10039    7.35000       (47.86667, 7.35)      True

#### Exercise 2.3
##### Using the meteorite data from the `Meteorite_Landings.csv` file, create a pivot table that shows both the number of meteorites and the 95th percentile of meteorite mass for those that were found versus observed falling per year from 2005 through 2009 (inclusive). Hint: Be sure to convert the `year` column to a number as we did in the previous exercise.

In [38]:
import pandas as pd
meteorites = pd.read_csv("../data/Meteorite_Landings.csv").assign(
    year=lambda x: pd.to_numeric(x.year.str.slice(6, 10))
).query(
 '2005 <= year <= 2009'
).pivot_table(
    index='year', columns='fall', aggfunc=['count', lambda x: x.quantile(0.95)], values='id'
).rename(columns={'<lambda>' : 'mass (95 percentile)'})
meteorites

count       mass (95 percentile)      
fall  Fell Found                 Fell Found
year                                       
2005   NaN   875                  NaN 49484
2006     5  2451                47360 50626
2007     8  1181                51805 53298
2008     9   948                52875 54119
2009     5  1492                51453 56009

#### Exercise 2.4
##### Using the meteorite data from the `Meteorite_Landings.csv` file, compare summary statistics of the mass column for the meteorites that were found versus observed falling.

#### Exercise 2.5
##### Using the taxi trip data in the `2019_Yellow_Taxi_Trip_Data.csv` file, resample the data to an hourly frequency based on the dropoff time. Calculate the total `trip_distance`, `fare_amount`, `tolls_amount`, and `tip_amount`, then find the 5 hours with the most tips.

--- 

### Section 3

#### Exercise 3.1
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create box plots for traveler throughput for each year in the data. Hint: Pass `kind='box'` into the `plot()` method to generate box plots.

#### Exercise 3.2
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create a heatmap that shows the 2019 TSA median traveler throughput by day of week and month.

#### Exercise 3.3
##### Annotate the medians in the box plot from *[Exercise 3.1](#Exercise-3.1)*. Hint: The `x` coordinates will be 1, 2, and 3 for 2019, 2020, and 2021, respectively. Alternatively, to avoid hardcoding values, you can use the `Axes.get_xticklabels()` method, in which case you should look at the [documentation](https://matplotlib.org/stable/api/text_api.html) for the `Text` class.